draft-ietf-mls-protocol.md

```diff
@@ -1910,46 +1911,45 @@ The client verifies the validity of a LeafNode using the following steps:
   another LeafNode, the authentication service MUST additionally validate that
   the set of identities attested by the credential in the new LeafNode is
   acceptable relative to the identities attested by the old credential.
+  E.g. sending an Update proposal updates the sender's old LeafNode to a new
+  one, or joining via external commit updates the joiner's old LeafNode
+  included in the committed Remove proposal to a new one.
 
-* Verify that the signature on the LeafNode is valid using the public key
-  in the LeafNode's credential
+* Verify that the signature on the LeafNode is valid using the LeafNode's
+  `signature_key`.
 
 * Verify that the LeafNode is compatible with the group's parameters.  If the
   GroupContext has a `required_capabilities` extension, then the required
-  extensions, proposals, and credential types MUST be listed in the LeafNode's `capabilities`
-  field.
+  extensions, proposals, and credential types MUST be listed in the LeafNode's
+  `capabilities` field.
 
 * Verify that the credential type is supported by all members of the group, as
   specified by the `capabilities` field of each member's LeafNode, and that the
   `capabilities` field of this LeafNode indicates support for all the credential
   types currently in use by other members.
 
 * Verify the `lifetime` field:
-  * When validating a LeafNode in a KeyPackage before sending an Add proposal,
-    the current time MUST be within the `lifetime` range.  A KeyPackage
-    containing a LeafNode that is expired or not yet valid MUST NOT be sent in
-    an Add proposal.
-  * When receiving an Add or validating a tree, checking the `lifetime` is
-    RECOMMENDED, if it is feasible in a given application context.  Because of
-    the asynchronous nature of MLS, the `lifetime` may have been valid when the
-    leaf node was proposed for addition, even if it is expired at these later
-    points in the protocol.
-
-* Verify that the `leaf_node_source` field has the appropriate value for the
-  context in which the LeafNode is being validated (as defined in
-  {{leaf-node-contents}}).
-
-* Verify that the following fields in the LeafNode are unique among the
-  members of the group (including any other members added in the same
-  Commit):
-
-    * `encryption_key`
-    * `signature_key`
```

Comment:
Why are you removing this uniqueness requirement?

Reply:
I see you have added it below.  That seems like a lot of repetition and risk of decoherence.  I would prefer to have it up here.

If the concern is the "added in the same Commit" part -- it seems like this requirement could be applied iteratively, in the sense that if you're updating/adding a bunch of leaves in a Commit and you do this check relative to the current tree before you apply an update/add, then that will catch a duplication.

Reply:
On second thought, I think this check belongs here.

Re. "added in the same Commit", I think a fix for this would be to change the "Validating Proposals" procedure to check individual validity at the end, after applying all committed proposals. I.e. the list is invalid if "After applying the proposal list, an individual proposal is invalid as specified in {{proposals}}."

Re. applying the check iteratively, the way I read the current draft, it seems that all checks are applied to the current state. For example, its valid to commit to A removing B and B removing A. But if we applied checks iteratively, it would be invalid.